File: src/ur5_pkg/ur5_kinematics/include/ur5_kinematics.cpp
#include "ur5_kinematics.h"

// ur5_model_param
#define D1 0.08935245198864346
#define D4 0.1108629663680774
#define D5 0.09481752424615827
#define D6 0.08250304262136304
#define A2 0.4255142946511841
#define A3 0.3922769788

Eigen::Matrix4d URposition::dhfwardkmatics(double joint_val[6])
{
    Eigen::Matrix4d tm_matrix;
    double tm_solution[4][4];
	double c1 = 0, s1 = 0, c2 = 0, s2 = 0, c5 = 0, s5 = 0, c6 = 0, s6 = 0, c23 = 0, s23 = 0, c234 = 0, s234 = 0;
	c1 = cos(joint_val[0]);
	s1 = sin(joint_val[0]);
	c2 = cos(joint_val[1]);
	s2 = sin(joint_val[1]);
	c5 = cos(joint_val[4]);
	s5 = sin(joint_val[4]);
	c6 = cos(joint_val[5]);
	s6 = sin(joint_val[5]);
	c23 = cos(joint_val[1] + joint_val[2]);
	s23 = sin(joint_val[1] + joint_val[2]);
	c234 = cos(joint_val[1] + joint_val[2] + joint_val[3]);
	s234 = sin(joint_val[1] + joint_val[2] + joint_val[3]);

	tm_solution[0][0] = c1 * c234*c5*c6 + s1 * s5*c6 - c1 * s234*s6;
	tm_solution[0][1] = -c1 * c234*c5*s6 - s1 * s5*s6 - c1 * s234*c6;
	tm_solution[0][2] = -c1 * c234*s5 + s1 * c5;
	tm_solution[0][3] = -c1 * c234*s5*D6 + s1 * c5*D6 + c1 * s234*D5 - c1 * (A3*c23 + A2 * c2) + D4 * s1;
	tm_solution[1][0] = s1 * c234*c5*c6 - c1 * s5*c6 - s1 * s234*s6;
	tm_solution[1][1] = -s1 * c234*c5*s6 + c1 * s5*s6 - s1 * s234*c6;
	tm_solution[1][2] = -s1 * c234*s5 - c1 * c5;
	tm_solution[1][3] = -s1 * c234*s5*D6 - c1 * c5*D6 + s1 * s234*D5 - s1 * (A3*c23 + A2 * c2) - D4 * c1;
	tm_solution[2][0] = s234 * c5*c6 + c234 * s6;
	tm_solution[2][1] = -s234 * c5*s6 + c234 * c6;
	tm_solution[2][2] = -s234 * s5;
	tm_solution[2][3] = -s234 * s5*D6 - c234 * D5 - (A3*s23 + A2 * s2) + D1;
	tm_solution[3][0] = 0;
	tm_solution[3][1] = 0;
	tm_solution[3][2] = 0;
	tm_solution[3][3] = 1;

    tm_matrix << tm_solution[0][0],tm_solution[0][1],tm_solution[0][2],tm_solution[0][3],
                 tm_solution[1][0],tm_solution[1][1],tm_solution[1][2],tm_solution[1][3],
                 tm_solution[2][0],tm_solution[2][1],tm_solution[2][2],tm_solution[2][3],
                 tm_solution[3][0],tm_solution[3][1],tm_solution[3][2],tm_solution[3][3];



	return tm_matrix;
}



bool robot_pose_check(double (&joint)[6])
{
  Eigen::Vector2d link2, link3, link5;
  link2 << A2 * cos(PI + joint[1]), A2 * sin(PI + joint[1]);
  link3 << A3 * cos(PI + joint[1] + joint[2]), A3 * sin(PI + joint[1] + joint[2]);
  link5 << D5 * cos(PI + joint[1] + joint[2] + PI/2 + joint[3]), D5 * sin(PI + joint[1] + joint[2] + PI/2 + joint[3]);

  //eblow up
  Eigen::Vector2d link23, link23_unit, link23_normal;
  link23 = link2 + link3;
  link23_unit = link23 / link23.norm();

  link23_normal = link2 - (link2.dot(link23_unit)) * link23_unit;
  if(link23_normal.dot(Eigen::Vector2d::UnitY())<0) return false;
  
/*   //wrist down
  Eigen::Vector2d link235;
  link235 = link2 + link3 + link5;
  if(link235.dot(Eigen::Vector2d::UnitY())<0) return false; */

  return true;
}

/*******************************************************************************/
//class Dist
double Dist::distance[6] = {0,0,0,0,0,0};

Dist::Dist(/* args */)
{
  step = 1.0*PI/180;
}

Dist::~Dist(){}

void Dist::euler_trans(Eigen::Vector3d base, Eigen::Vector3d &euler)
{
  //goal(0) in (0,pi)
  if(fabs(euler(0)-base(0)) > (PI/2)){
    if(euler(0) < base(0)) euler(0) += PI;
    else euler(0) -= PI;

    double e0_ = PI - euler[1];
    euler[1]  = e0_;
    euler[2] -= PI;
  }
  shiftangle(euler(1),base(1));
  shiftangle(euler(2),base(2));
  return;
}

void Dist::trans(EIso3d pose_goal, EIso3d pose_now)
{
  Eigen::Vector3d r_goal = pose_goal.rotation().eulerAngles(0,1,2);
  Eigen::Vector3d r_now = pose_now.rotation().eulerAngles(0,1,2);
  //std::cout << "Goal" << r_goal.transpose() << std::endl;
  //std::cout << "Now" << r_now.transpose() << std::endl;

  euler_trans(r_now, r_goal);
  //std::cout << "Goal" << r_goal.transpose() << std::endl;

  for(unsigned long i=0;i<3;i++){
    p(i) = pose_goal.translation()(i) - pose_now.translation()(i);
    r(i) = r_goal(i) - r_now(i);
    distance[i]   = p(i);
    distance[3+i] = r(i);
  }
  return;
}

double Dist::lenth()
{
  pose.block<3,1>(0,0) = p * 1000;
  pose.block<3,1>(3,0) = r * 180.0/PI;
  return pose.norm();
}

EVec6d Dist::delta_pose()
{
  pose.block<3,1>(0,0) = p;
  pose.block<3,1>(3,0) = r;
  return pose;
}

/*******************************************************************************/
//URposition
URposition::URposition(){}

URposition::~URposition(){}

double URposition::jointbase[6] = {0,-PI/2,0,0,0,0};
EMat26d URposition::joint_limit = (EMat26d() << -PI*2, -PI, -PI*5/6, -PI*4/3, -PI*2, -PI*2,
                                                                                     PI*2,   0,  PI*5/6,    PI/3,  PI*2,  PI*2).finished();

void URposition::load_limit(EMat26d limit_in)
{
  joint_limit = limit_in;
}

void URposition::set_jointbase(double jointval_in[6])
{
	for(int i=0; i<6; i++) jointbase[i] = jointval_in[i];
}


bool URposition::inv_solve(EIso3d &mat, double (&joint)[6], int (&flag_in)[3], double joint6)
{
	// theta1
  double t1_ = pow(mat(0,3) - D6 * mat(0,2), 2) + pow(-mat(1,3) + D6 * mat(1,2), 2) - pow(D4, 2);
	if (t1_ < 0) {
		std::cout << "机械臂无法达到该范围,(x,y)需要远离基座标系原点！\n";
		return false;
	}
	joint[0] = atan2(D4, flag_in[0]*sqrt(t1_)) - atan2(-mat(1,3) + D6 * mat(1,2), mat(0,3) - D6 * mat(0,2));
	limitangle(joint[0]);
	double tc1 = 0;
	double ts1 = 0;
	tc1 = cos(joint[0]);
	ts1 = sin(joint[0]);

	// theta5
  double t5_ = ts1 * mat(0,2) - tc1 * mat(1,2);
	if (pow(t5_,2) > 1) {
		std::cout << "Rotation matrix is not qualified\n";
		return false;
	}
  joint[4] = atan2(flag_in[1]*sqrt(1.0 - pow(t5_,2)), t5_);
	limitangle(joint[4]);

	// theta6
	int flag6 = sign(sin(joint[4]));
  double theta234 = 0;
  if(flag6 == 0){
    joint[5] = joint6;
    limitangle(joint[5]);
    theta234 = atan2(mat(2,0), tc1*mat(0,0) + ts1 * mat(1,0)) - joint[5];
  }
  else{
    joint[5] = atan2(flag6*(-ts1 * mat(0,1) + tc1 * mat(1,1)), flag6*(ts1*mat(0,0) - tc1 * mat(1,0)));
	  limitangle(joint[5]);
    theta234 = atan2(-mat(2,2) * flag6, -(tc1*mat(0,2) + ts1 * mat(1,2))*flag6);
  }

	double ts5 = sin(joint[4]);
	double tc234 = cos(theta234);
	double ts234 = sin(theta234);

	// theta2
	double M = -tc1 * mat(0,3) - ts1 * mat(1,3) - tc234 * ts5*D6 + ts234 * D5;
	double N = D1 - mat(2,3) - ts234 * ts5*D6 - tc234 * D5;
	double mn = M * M + N * N + A2 * A2 - A3 * A3;

	double t2_ = 4.0 * A2*A2 * (M*M + N*N) - pow(mn, 2);
	if (t2_ < 0) {
		std::cout << "机械臂无法达到该范围，(x,y,z)不能超过半径750mm的球型空间!\n";
		return false;
	}
	joint[1] = atan2(mn, flag_in[2] * sqrt(t2_)) - atan2(M*A2, N*A2);
	limitangle(joint[1]);

	// theta3
	double tc2 = cos(joint[1]);
	double ts2 = sin(joint[1]);

	joint[2] = atan2(N - A2 * ts2, M - A2 * tc2) - joint[1];
	limitangle(joint[2]);

	// theta4
	joint[3] = theta234 - joint[1] - joint[2];
	limitangle(joint[3]);

	return true;
}



bool URposition::dhinvkinematics(EIso3d &mat, double (&joint)[6], double (&joint_base)[6])
{
	int flag[8][3] = { {1,1,1}, {1,1,-1}, {1,-1,1}, {1,-1,-1}, {-1,1,1}, {-1,1,-1}, {-1,-1,1}, {-1,-1,-1} };

  double solution[8][6];
	bool joint_val_check[8];
	bool has_solution = false;

	for(int i=0; i<8; i++){
		if(inv_solve(mat, solution[i], flag[i], joint_base[5])){
      has_solution = true;
      shiftangle(solution[i],joint_base);
			joint_val_check[i] = check_joint(joint_limit, solution[i]);
      //printf solution
      if(false){
        std::cout << "fail one solution:" << std::endl;
        std::cout << solution[i][0]*180/PI << ", " << solution[i][1]*180/PI << ", " << solution[i][2]*180/PI << ", "  
                  << solution[i][3]*180/PI << ", " << solution[i][4]*180/PI << ", " << solution[i][5]*180/PI << std::endl;
      }
		}
		else{
			joint_val_check[i] = false;
		}
	}

	if(!has_solution){
		printf("Solutions all failed\n");
		return false;
	}

  //check
  for(int i=0; i<8; i++) if(joint_val_check[i]) if(!robot_pose_check(solution[i])) {joint_val_check[i] = false;
  if(false){
    std::cout << "fail one solution plus:" << std::endl;
    std::cout << solution[i][0]*180/PI << ", " << solution[i][1]*180/PI << ", " << solution[i][2]*180/PI << ", "  
              << solution[i][3]*180/PI << ", " << solution[i][4]*180/PI << ", " << solution[i][5]*180/PI << std::endl;
    }
  }
      

	// select solution
	int choice = 8;
  double distance = 12*PI;
  for(int i=0; i<8; i++){
    if(joint_val_check[i]){ 
      double dis_sum = 0;
      for(unsigned long j=0;j<6;j++) dis_sum += fabs(solution[i][j] - joint_base[j]);
      if(distance > dis_sum){
        distance = dis_sum;
        choice = i;
      }
    }
  }

	if(choice>=0 && choice<=7){
		for(int i=0; i<6; i++) joint[i] = solution[choice][i];
		//printf("Adopted Inverse kinematics solution: [%6.4f %6.4f %6.4f %6.4f %6.4f %6.4f]\n\n",joint[0],joint[1],joint[2],joint[3],joint[4],joint[5]);
		return true;
	}else{
		//printf("failed\n\n");
		return false;
	}
}

void URposition::select_mat(EIso3d &mat, int type)
{
	Eigen::Matrix3d rotate = Eigen::Matrix3d::Identity();
    Eigen::Vector3d point;
    Eigen::AngleAxisd ang;
	Eigen::Vector3d eul;
	switch (type)
    {
	//角轴表示
    case 1:
    {
      std::cout << "角轴表示输入格式:" << std::endl;
      std::cout << "角度(°) 轴X(m) 轴Y(m) 轴Z(m) 点X(m) 点Y(m) 点Z(m)" << std::endl;
      std::cin >> ang.angle() >> ang.axis()[0] >> ang.axis()[1] >> ang.axis()[2] >> point(0) >> point(1) >> point(2); 
      ang.angle() = ang.angle()/180*PI;

      rotate = ang.matrix();

	  // 设置旋转矩阵
	  mat.linear() = rotate;
	  // 设置平移向量
	  mat.translation() = Eigen::Vector3d(point[0], point[1], point[2]);

      break;
    }
	//欧拉角表示
    case 2:
    {
      std::cout << "欧拉角输入格式：" << std::endl;
      std::cout << "角度1(°) 角度2(°) 角度3(°) 点X(m) 点Y(m) 点Z(m)" << std::endl;
      std::cin >> eul(0) >> eul(1) >> eul(2) >> point(0) >> point(1) >> point(2); 
      eul = eul/180*PI;
      Eigen::AngleAxisd eul_x(Eigen::AngleAxisd(eul(0),Eigen::Vector3d::UnitX()));
      Eigen::AngleAxisd eul_y(Eigen::AngleAxisd(eul(1),Eigen::Vector3d::UnitY()));
      Eigen::AngleAxisd eul_z(Eigen::AngleAxisd(eul(2),Eigen::Vector3d::UnitZ()));
      ang = eul_x * eul_y * eul_z;
      Eigen::Vector3d eul_2;

      rotate = ang.matrix();
	  // 设置旋转矩阵
	  mat.linear() = rotate;
	  // 设置平移向量
	  mat.translation() = Eigen::Vector3d(point[0], point[1], point[2]);

      break;
    }
	//关节表示
	//-10 -80 -100 -150 -100 -100
	//-10 -80 -100 -10 90 90
    case 3:
    {
      double joint6[6];
      std::cout << "关节角度输入格式：" << std::endl;
      std::cout << "关节1(°) 关节2(°) 关节3(°) 关节4(°) 关节5(°) 关节6(°)" << std::endl;
      std::cin >> joint6[0] >> joint6[1] >> joint6[2] >> joint6[3] >> joint6[4] >> joint6[5]; 
      for(unsigned long i=0;i<6;i++) joint6[i] = joint6[i]/180*PI;

	  Eigen::Matrix4d tm_matrix = dhfwardkmatics(joint6);
	  mat = tm_matrix;

      break;
    }

	//四元数
	//0.2 0.2 0.3 0.1 0.1 0.1 0.1 
	//0.4 0.1 0.3 0.2 0.2 0.2 0.1
	case 4:
	{
		// 初始化一个数组用于存储输入的位置向量
		double position[3];
		// 初始化一个数组用于存储输入的四元数
		double quaternion[4];
		// 提示用户输入格式
		std::cout << "输入格式为:" << std::endl;
		std::cout << "位置X(m) 位置Y(m) 位置Z(m) 四元数w 四元数x 四元数y 四元数z" << std::endl;
		// 接收用户输入的位置向量和四元数
		std::cin >> position[0] >> position[1] >> position[2] 
				>> quaternion[0] >> quaternion[1] >> quaternion[2] >> quaternion[3];
		
		
		// 创建四元数
		Eigen::Quaterniond q(quaternion[0], quaternion[1], quaternion[2], quaternion[3]);
		// 正规化四元数，以防输入未正规化
		q.normalize();
		// 将四元数转换为旋转矩阵
		Eigen::Matrix3d m_rot = q.toRotationMatrix();

		// 设置旋转矩阵
		mat.linear() = m_rot;
		// 设置平移向量
		mat.translation() = Eigen::Vector3d(position[0], position[1], position[2]);

		break;
	}

	}

}

File: src/ur5_pkg/ur5_control/ur5_control/ur5_kinematics.py
import math
import numpy as np
cos = math.cos
sin = math.sin
PI = math.pi

D1 = 0.08935245198864346
D4 = 0.1108629663680774
D5 = 0.09481752424615827
D6 = 0.08250304262136304
A2 = 0.4255142946511841
A3 = 0.3922769788

class URposition:
    """
    求解UR运动学的类
    """
    def __init__(self) -> None:
        self.joint_limit = np.array([[-PI*2, -PI, -PI*5/6, -PI*4/3, -PI*2, -PI*2],
                                     [PI*2,   0,  PI*5/6,    PI/3,  PI*2,  PI*2]])
        self.joint_ref = np.array([0,0,0,0,0,0])
        pass

    def set_joint_ref(self, joint_set: np.ndarray):
        """
        设置参考关节坐标
        """
        assert isinstance(joint_set, np.ndarray), "Input must be a numpy array"
        assert joint_set.ndim == 1 and joint_set.shape[0] == 6, "Input must be a 1D array with 6 elements"
        self.joint_ref = joint_set

    def fwdkinematics_dh(self, joint_val: np.ndarray) -> np.ndarray:
        """
        计算正运动学(DH建模法)
        """
        c1 = cos(joint_val[0])
        s1 = sin(joint_val[0])
        c2 = cos(joint_val[1])
        s2 = sin(joint_val[1])
        c5 = cos(joint_val[4])
        s5 = sin(joint_val[4])
        c6 = cos(joint_val[5])
        s6 = sin(joint_val[5])
        c23 = cos(joint_val[1] + joint_val[2])
        s23 = sin(joint_val[1] + joint_val[2])
        c234 = cos(joint_val[1] + joint_val[2] + joint_val[3])
        s234 = sin(joint_val[1] + joint_val[2] + joint_val[3])

        tm = np.eye(4)
        tm[0][0] = c1 * c234 * c5 * c6 + s1 * s5 * c6 - c1 * s234 * s6
        tm[0][1] = -c1 * c234 * c5 * s6 - s1 * s5 * s6 - c1 * s234 * c6
        tm[0][2] = -c1 * c234 * s5 + s1 * c5
        tm[0][3] = -c1 * c234 * s5 * D6 + s1 * c5 * D6 + c1 * s234 * D5 - c1 * (A3 * c23 + A2 * c2) + D4 * s1
        tm[1][0] = s1 * c234 * c5 * c6 - c1 * s5 * c6 - s1 * s234 * s6
        tm[1][1] = -s1 * c234 * c5 * s6 + c1 * s5 * s6 - s1 * s234 * c6
        tm[1][2] = -s1 * c234 * s5 - c1 * c5
        tm[1][3] = -s1 * c234 * s5 * D6 - c1 * c5 * D6 + s1 * s234 * D5 - s1 * (A3 * c23 + A2 * c2) - D4 * c1
        tm[2][0] = s234 * c5 * c6 + c234 * s6
        tm[2][1] = -s234 * c5 * s6 + c234 * c6
        tm[2][2] = -s234 * s5
        tm[2][3] = -s234 * s5 * D6 - c234 * D5 - (A3 * s23 + A2 * s2) + D1

        return tm

    def invkinematics_dh(self, mat: np.ndarray, joint: np.ndarray) -> bool:
        """
        计算逆运动学(DH建模法)
        """
        flag = np.array([[1, 1, 1], [1, 1, -1], [1, -1, 1], [1, -1, -1],
                         [-1, 1, 1], [-1, 1, -1], [-1, -1, 1], [-1, -1, -1]])

        solution = np.zeros((8, 6), dtype=np.float64)
        joint_val_check = np.zeros(8)
        has = False

        for i in range(8):
            if self.inv_solve(mat, solution[i], flag[i], self.joint_ref[5]):
                has = True
                shiftangle_array(solution[i],self.joint_ref)
                joint_val_check[i] = check_joint_array(self.joint_limit, solution[i])
                # Print solution
                if False:
                    print("fail one solution:")
                    print(solution[i] * 180 / math.pi)

            else:
                joint_val_check[i] = False

        if not has:
            print("Solutions all failed")
            return False

        # Check
        for i in range(8):
            if joint_val_check[i]:
                if not robot_pose_check(solution[i]):
                    joint_val_check[i] = False
        if False:
            print("fail one solution plus:")
            print(solution[i] * 180 / math.pi)

        # Select solution
        choice = 8
        distance = 12 * math.pi
        for i in range(8):
            if joint_val_check[i]:
                dis_sum = np.sum(np.abs(solution[i] - self.joint_ref))
                if distance > dis_sum:
                    distance = dis_sum
                    choice = i

        if 0 <= choice <= 7:
            joint[:] = solution[choice]
            # print("Adopted Inverse kinematics solution:", joint)
            return True
        else:
            # print("failed")
            return False

    def inv_solve(self, mat, joint, flag_in, joint6):
        """
        计算指定的一组逆运动学解(DH建模法)
        """
        # theta1
        t1_ = (mat[0, 3] - D6 * mat[0, 2]) ** 2 + (-mat[1, 3] + D6 * mat[1, 2]) ** 2 - D4 ** 2
        if t1_ < 0:
            print("机械臂无法达到该范围,(x,y)需要远离基座标系原点！")
            return False
        joint[0] = np.arctan2(D4, flag_in[0] * np.sqrt(t1_)) - np.arctan2(-mat[1, 3] + D6 * mat[1, 2],
                                                                          mat[0, 3] - D6 * mat[0, 2])
        limitangle(joint[0])
        tc1 = cos(joint[0])
        ts1 = sin(joint[0])

        # theta5
        t5_ = ts1 * mat[0, 2] - tc1 * mat[1, 2]
        if t5_ ** 2 > 1:
            print("Rotation matrix is not qualified")
            return False
        joint[4] = np.arctan2(flag_in[1] * np.sqrt(1.0 - t5_ ** 2), t5_)
        joint[4] = limitangle(joint[4])

        # theta6
        flag6 = np.sign(np.sin(joint[4]))
        if flag6 == 0:
            joint[5] = joint6
            joint[5] = limitangle(joint[5])
            theta234 = np.arctan2(mat[2, 0], tc1 * mat[0, 0] + ts1 * mat[1, 0]) - joint[5]
        else:
            joint[5] = np.arctan2(flag6 * (-ts1 * mat[0, 1] + tc1 * mat[1, 1]),
                                  flag6 * (ts1 * mat[0, 0] - tc1 * mat[1, 0]))
            joint[5] = limitangle(joint[5])
            theta234 = np.arctan2(-mat[2, 2] * flag6, -(tc1 * mat[0, 2] + ts1 * mat[1, 2]) * flag6)

        ts5 = sin(joint[4])
        tc234 = cos(theta234)
        ts234 = sin(theta234)

        # theta2
        M = -tc1 * mat[0, 3] - ts1 * mat[1, 3] - tc234 * ts5 * D6 + ts234 * D5
        N = D1 - mat[2, 3] - ts234 * ts5 * D6 - tc234 * D5
        mn = M * M + N * N + A2 * A2 - A3 * A3

        t2_ = 4.0 * A2 * A2 * (M * M + N * N) - mn ** 2
        if t2_ < 0:
            print("机械臂无法达到该范围，(x,y,z)不能超过半径750mm的球型空间!")
            return False
        joint[1] = np.arctan2(mn, flag_in[2] * np.sqrt(t2_)) - np.arctan2(M * A2, N * A2)
        joint[1] = limitangle(joint[1])

        # theta3
        tc2 = cos(joint[1])
        ts2 = sin(joint[1])

        joint[2] = np.arctan2(N - A2 * ts2, M - A2 * tc2) - joint[1]
        joint[2] = limitangle(joint[2])

        # theta4
        joint[3] = theta234 - joint[1] - joint[2]
        joint[3] = limitangle(joint[3])

        return True


def limitangle(theta):
    while theta > PI:
        theta -= 2.0 * PI
    while theta <= (-PI + 0.0001):
        theta += 2.0 * PI
    return theta


def limitangle_array(theta):
    for i in range(len(theta)):
        theta[i] = limitangle(theta[i])


def shiftangle(theta, base):
    while theta > (base + PI):
        theta -= 2.0 * PI
    while theta < (base - PI + 0.0001):
        theta += 2.0 * PI
    return theta


def shiftangle_array(theta, base):
    for i in range(len(base)):
        theta[i] = shiftangle(theta[i], base[i])


def check_joint(limit, index, joint):
    if joint < limit[0, index]:
        return False
    if joint > limit[1, index]:
        return False
    return True


def check_joint_array(limit, joint):
    for i in range(len(joint)):
        val = joint[i]
        while val <= limit[0, i]:
            val += PI * 2.0
        while val >= limit[1, i]:
            val -= PI * 2.0
        if val <= limit[0, i]:
            return False
        else:
            joint[i] = val
    return True


def robot_pose_check(joint):

    link2 = np.array([A2 * np.cos(np.pi + joint[1]), A2 * np.sin(np.pi + joint[1])])
    link3 = np.array([A3 * np.cos(np.pi + joint[1] + joint[2]), A3 * np.sin(np.pi + joint[1] + joint[2])])
    link5 = np.array([D5 * np.cos(np.pi + joint[1] + joint[2] + np.pi / 2 + joint[3]),
                      D5 * np.sin(np.pi + joint[1] + joint[2] + np.pi / 2 + joint[3])])

    link23 = link2 + link3
    link23_unit = link23 / np.linalg.norm(link23)

    link23_normal = link2 - np.dot(link2, link23_unit) * link23_unit
    if link23_normal.dot(np.array([0, 1])) < 0:
        return False

    return True


def pose_to_matrix(pose):
    X, Y, Z, RX, RY, RZ = pose

    # Rotation matrices for RX (Roll), RY (Pitch), RZ (Yaw)
    R_X = np.array([[1, 0, 0],
                    [0, math.cos(RX), -math.sin(RX)],
                    [0, math.sin(RX), math.cos(RX)]])

    R_Y = np.array([[math.cos(RY), 0, math.sin(RY)],
                    [0, 1, 0],
                    [-math.sin(RY), 0, math.cos(RY)]])

    R_Z = np.array([[math.cos(RZ), -math.sin(RZ), 0],
                    [math.sin(RZ), math.cos(RZ), 0],
                    [0, 0, 1]])

    # Combined rotation matrix
    R = np.dot(R_Z, np.dot(R_Y, R_X))

    # Translation vector
    t = np.array([X, Y, Z])

    # Construct the pose matrix (4x4 homogeneous transformation matrix)
    pose_matrix = np.eye(4)
    pose_matrix[:3, :3] = R
    pose_matrix[:3, 3] = t

    return pose_matrix

def matrix_to_pose(mat):
    # Extract rotation matrix and translation vector
    R = mat[:3, :3]
    t = mat[:3, 3]

    # Calculate Roll (RX), Pitch (RY), Yaw (RZ) using Euler angles (RPY)
    RX = math.atan2(R[2, 1], R[2, 2])
    RY = math.atan2(-R[2, 0], math.sqrt(R[2, 1]**2 + R[2, 2]**2))
    RZ = math.atan2(R[1, 0], R[0, 0])

    # Calculate X, Y, Z positions
    X = t[0]
    Y = t[1]
    Z = t[2]

    # Return [X, Y, Z, RX, RY, RZ]
    return [X, Y, Z, RX, RY, RZ]

# def RPY_rad_to_deg(RPY):
#     X, Y, Z, RX, RY, RZ = RPY
#     RX = RX*180/PI
#     RY = RY*180/PI
#     RZ = RZ*180/PI
#     return [X, Y, Z, RX, RY, RZ]

def joint_deg_to_rad(joint_deg):
    joint_rad = [0]*6
    for i in range(6):
        joint_rad[i] = joint_deg[i]*PI/180
    return joint_rad

"""
UR5 kinematics Demo(FOR TEST)
"""

def main(args=None):
    a = URposition()
    # l1 = np.array([0, -PI/2, -PI/2, -PI/2, PI/2, 0])
    l1 = np.array([21.15, -94.59, -110.76, -156.41, 286.34, -91.47])
    l1 = joint_deg_to_rad(l1)
    print("l1关节表示：\n")
    print(l1)
    l2 = a.fwdkinematics_dh(l1)
    print("l1正运动学求出矩阵：\n")
    print(l2)
    print("矩阵转为RPY表示：\n")
    print(matrix_to_pose(l2))
    joint = l1
    a.invkinematics_dh(l2, joint)
    print("矩阵逆运动求解出关节：\n")
    print(joint)
    print("---------------------------------\n")

    # RPY = np.array([0.48740,-0.10922,0.43196,2.223,-2.224,-0.004])
    RPY = np.array([0.48144,0.04405,0.24782,0.267,-4.672,-0.115])
    print("RPY表示：\n")
    print(RPY)
    print("RPY转为旋转矩阵表示：\n")
    print(pose_to_matrix(RPY))
    a.invkinematics_dh(pose_to_matrix(RPY), joint)
    print("RPY矩阵逆运动求解出关节：\n")
    print(joint)

if __name__ == '__main__':
    main()